src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/DependenciesChangesBuilder.cs

```diff
@@ -3,58 +3,56 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Linq;
-
 using Microsoft.VisualStudio.Imaging.Interop;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.Subscriptions
 {
-    internal sealed class DependenciesChanges : IDependenciesChanges
+    internal sealed class DependenciesChangesBuilder
     {
-        private readonly HashSet<IDependencyModel> _added = new HashSet<IDependencyModel>();
-        private readonly HashSet<IDependencyModel> _removed = new HashSet<IDependencyModel>();
-
-        public bool AnyChanges => _added.Count != 0 || _removed.Count != 0;
+        private HashSet<IDependencyModel> _added;
+        private HashSet<IDependencyModel> _removed;
 
-        public IImmutableList<IDependencyModel> AddedNodes
+        public void Added(IDependencyModel model)
         {
-            get
+            if (_added == null)
             {
-                lock (_added)
-                {
-                    return ImmutableArray.CreateRange(_added);
-                }
+                _added = new HashSet<IDependencyModel>();
             }
+
+            _added.Remove(model);
+            _added.Add(model);
```

Comment:
I trust that you're really smart and stuff, so can you explain what this is for? Different IDependencyModel objects that have the same hashcode?

Reply:
This pattern's been there for a while, I've just moved it. It looks weird, but is necessary because `Add` is a no-op if the item's already present.

There's currently an implicit requirement that an `IDependencyModel` be equatable with others of its own kind. Formalising this is on my radar, but it's part of the public API so it might not be possible or practical to do so.

Reply:
> is necessary because Add is a no-op if the item's already present.

You say that like its a bad thing....

Reply:
You prompted me to look at this again.

`DependencyRulesSubscriber` has a list of `IDependenciesRuleHandler` implementations, which it goes through in a precedence order in response to `IProjectChangeDescription`s.

In theory two different handlers could produce different versions of the same dependency. I don't know if this happens, or if there's a reason to block it. However in this case it's not part of the public API so it's probably safe to change, maybe throwing if `Add` returns `false`.

If it meant we could get rid of the equatability requirement on `IDependency` it might be worth the work.